tools/ratbagctl.body.py.in

```diff
@@ -0,0 +1,1889 @@
+name = "name"
+"""the group to logically associate commands while printing the help"""
+group = "group"
+"""list of positional arguments for the given command"""
+pos_args = "pos_args"
+"""a tag that we can refer latrer in an element of type 'link'"""
+tag = "tag"
+"""the element pointed to in an element of type 'link'"""
+dest = "dest"
+"""the function to associate to the switch or command"""
+func = "func"
+"""this is a particular command that is an integer, but not an terminating argument.
+example:
+ profile active get
+ profile **2** button 3 get
+ - "profile" needs to be a switch
+ - "2" needs to be translated as a N_access, given it is a requirement to be able to call 'button'
+ """
+N_access = "N_access"
+
+# argparse.add_argument parameters (forwarded as such)
+"""'type' of the argument"""
+arg_type = "arg_type"
+"""'metavar' of the argument"""
+metavar = "metavar"
+"""'help' of the argument"""
+help_str = "help"
+"""'nargs' of the argument"""
+nargs = "nargs"
+"""'choices' of the argument"""
+choices = "choices"
+
+# types
+"""an option to interpret as a command (example 'list', 'info')"""
+command = "command"
+"""an argument that is required for the given command arguments are leaf nodes
+and can not have children
+"""
+argument = "argument"
+"""provides a list of choice of commands for instance, a switch of [A, B] means
+we can have A or B only when parsing the command line
+"""
+switch = "switch"
+"""same as list, except we can loop inside the list for instance, a set of
+[A, B] means we can have A and B (and A, ...) one after the other, no matter
+the order
+"""
+set = "set"
+"""a reference to any other element in the tree marked with a tag"""
+link = "link"
+
+################################################################################
+
+
+def commit(device: RatbagdDevice, args: argparse.Namespace) -> None:
+    if args.nocommit:
+        return
+    device.commit()
+
+
+def color(string: str) -> Tuple[int, int, int]:
+    try:
+        int_value = int(string, 16)
+    except ValueError as e:
+        msg = "%r is not a color in hex format" % string
+        raise argparse.ArgumentTypeError(msg) from e
+    r = (int_value >> 16) & 0xFF
+    g = (int_value >> 8) & 0xFF
+    b = (int_value >> 0) & 0xFF
+    return (r, g, b)
+
+
+def u8(string: str) -> int:
+    int_value = int(string)
+    msg = "%r is not a single byte" % string
+    if int_value < 0 or int_value > 255:
+        raise argparse.ArgumentTypeError(msg)
+    return int_value
+
+
+def dpi(string: str) -> Union[Tuple[int, int], Tuple[int,]]:
+    try:
+        int_value = int(string)
+    except ValueError:
+        pass
+    else:
+        return (int_value,)
+    if string.endswith("dpi"):
+        string = string[:-3]
+    x, y = string.split("x")
+    try:
+        int_x = int(x)
+        int_y = int(y)
+    except ValueError as e:
+        raise argparse.ArgumentTypeError("%r is not a valid dpi" % string) from e
+    else:
+        return (int_x, int_y)
```

Comment:
nitpick: there's no need for the `else` block here, you can return right away

Reply:
Oh I hate necessary else statements, thanks! I didn't notice this because I only skimmed fast through changed lines to make sure there are no obvious breakages, but didn't actually spend time analyzing every single line.